csharp/src/Ice/Connection.cs

```diff
@@ -1658,4 +1609,158 @@ private enum State
+                // Cache the address to be able to access it after the socket is dispose
+                if (_remoteAddress == null)
+                {
+                    try
+                    {
+                        _remoteAddress = Transceiver.Fd()?.RemoteEndPoint as System.Net.IPEndPoint;
+                    }
+                    catch
+                    {
+                    }
+                }
+                return _remoteAddress;
+            }
+        }
+
+        private System.Net.IPEndPoint? _localAddress;
+        private System.Net.IPEndPoint? _remoteAddress;
+
+        public IPConnection(
+            Communicator communicator,
+            IACMMonitor? monitor,
+            ITransceiver transceiver,
+            IConnector? connector,
+            Endpoint endpoint,
+            ObjectAdapter? adapter)
+            : base(communicator, monitor, transceiver, connector, endpoint, adapter)
+        {
+        }
+    }
+
+    public class TcpConnection : IPConnection
+    {
+        public TcpConnection(
+            Communicator communicator,
+            IACMMonitor? monitor,
+            ITransceiver transceiver,
+            IConnector? connector,
+            Endpoint endpoint,
+            ObjectAdapter? adapter)
+            : base(communicator, monitor, transceiver, connector, endpoint, adapter)
+        {
+        }
+    }
+
+    public class UdpConnection : IPConnection
+    {
+        public System.Net.IPEndPoint? McastAddress
+        {
+            get
+            {
+                // Cache the address to be able to access it after the socket is dispose
+                if (_mcastAddress == null)
+                {
+                    try
+                    {
+                        _mcastAddress = (Transceiver as UdpTransceiver)?.McastAddress;
+                    }
+                    catch
+                    {
+                    }
+                }
+                return _mcastAddress;
+            }
+        }
+
+        private System.Net.IPEndPoint? _mcastAddress;
+
+        public UdpConnection(
+            Communicator communicator,
+            IACMMonitor? monitor,
+            ITransceiver transceiver,
+            IConnector? connector,
+            Endpoint endpoint,
+            ObjectAdapter? adapter)
+            : base(communicator, monitor, transceiver, connector, endpoint, adapter)
+        {
+        }
+    }
+
+    public class WSConnection : TcpConnection
+    {
+        public IReadOnlyDictionary<string, string> Headers => ((WSTransceiver)Transceiver).Headers;
+
+        public WSConnection(
+            Communicator communicator,
+            IACMMonitor? monitor,
+            ITransceiver transceiver,
+            IConnector? connector,
+            Endpoint endpoint,
+            ObjectAdapter? adapter)
+            : base(communicator, monitor, transceiver, connector, endpoint, adapter)
+        {
+        }
+    }
+
+    public class SslConnection : TcpConnection
```

Comment:
I think Tcp and Ssl should be merged in one class (TcpConnection); likewise, WS and Wss should be merged into WSConnection. But this can wait until a future PR.

Reply:
Sounds fine, I was thinking to do that, how will you check you are using a secure connection? seems to me that  `connection.Endpoint.IsSecure` should be enough, and don't need an extra `IsSecure` in the connection.

Reply:
fixed

Reply:
In a future PR, we should actually decouple "secure-capable endpoints" (with ice1, all "s" endpoints and with ice2, all endpoints) and "secure connections".